addons/account/models/account_move.py

```diff
@@ -3715,14 +3715,19 @@ def _check_constrains_account_id_journal_id(self):
 
     @api.constrains('account_id', 'tax_ids', 'tax_line_id', 'reconciled')
     def _check_off_balance(self):
+        checked_moves = set()
         for line in self:
-            if line.account_id.internal_group == 'off_balance':
-                if any(a.internal_group != line.account_id.internal_group for a in line.move_id.line_ids.account_id):
-                    raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
-                if line.tax_ids or line.tax_line_id:
-                    raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
-                if line.reconciled:
-                    raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
+            if line.account_id.internal_group != 'off_balance':
+                continue
+            if line.move_id.id in checked_moves:
+                continue
+            checked_moves.add(line.move_id.id)
+            if any(a.internal_group != 'off_balance' for a in line.move_id.line_ids.account_id):
+                raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
+            if any(True for l in line.move_id.line_ids if l.tax_ids or l.tax_line_id):
+                raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
+            if any(True for l in line.move_id.line_ids if l.reconciled):
+                raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
```

Comment:
I find this much more understandable and it should have the same complexity. The two `continue` clause are messy IMO and what about those `any(True for x in y if cond)`. Why not simply `any(cond for x in y)`?

By the way, there was a bug in the original code, so thanks for making me dive into it: 
* Create an off balance move
* Save it. 
* Change one line to a balance sheet account
* Result: No constraint raised because it didn't enter the first `if` clause.

```suggestion
        types_per_move = {move: set(move.line_ids.account_id.mapped('internal_group')) for move in self.move_id}
        for types in types_per_move.values():
            if 'off_balance' in types and len(types) != 1:
                raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))

        for line in self:
            if line.account_id.internal_group == 'off_balance':
                if line.tax_ids or line.tax_line_id:
                    raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
                if line.reconciled:
                    raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
```

Reply:
I have gone through similar approach while reviewing your comments.

What about this approach:

```
  @api.constrains('account_id', 'tax_ids', 'tax_line_id', 'reconciled')
  def _check_off_balance(self):
      off_moves = []
      checked_moves = set()
      for line in self:
          move = line.move_id
          move_id = move.id
          if move_id in checked_moves:
              continue
          checked_moves.add(move_id)
          internal_group = set([a.internal_group for a in move.line_ids.account_id])
          if 'off_balance' in internal_group and len(internal_group) > 1:
              raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
          if any(True for l in move.line_ids if l.tax_ids or l.tax_line_id):
              raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
          if any(True for l in move.line_ids if l.reconciled):
              raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
```

I use 
`any(True for x in y if cond)` rather than `any(cond for x in y)`
because when timing it. it is faster.

Regards.